relay.go

```diff
@@ -390,7 +400,10 @@ func (r *Relay) handleHopStream(s inet.Stream, msg *pb.CircuitRelay) {
 	}()
 
 	go func() {
-		count, err := io.Copy(bs, s)
+		buf := r.bufPool.Get().([]byte)
+		defer r.bufPool.Put(buf)
```

Comment:
Can we zero out the buffer before putting it back? It might not be necessary in practice, but would rather do it so getting a buffer from the pool always returns a zero value byte slice, whether new or pooled.

Reply:
I don't think it's necessary, as it will contain encrypted data. Also, zeroing out takes time, I'd rather keep it as lightweight as possible.

Reply:
Also, the way `CopyBuffer` is implemented it doesn't look at the data. At worst we might leak some old encrypted data to the reader, which a) isn't happening in our stack and b) it is just random junk thanks to the encryption.

Reply:
I think we should do it. Lets not leak data even if it’s encrypted. Leaking data would also corrupt the SecIO/TLS ciphers. Zeroing an array is very cheap.

Reply:
sigh.... I think it's completely unnecessary, but ok.

Reply:
There is no utility function to do that in go, and I really hate to have to write bzero.
Also, it's an O(n) operation.

Reply:
Also, how on earth are we going to leak data with it? 
The reader just puts bytes in the slice and returns the number of bytes read.
The writer then writes only the bytes that have been read.

Reply:
Also of note that we started with the intention of making it more efficient than allocating a new buffer. If we add zeroing out we pretty much lost all efficiency gains.

Reply:
> The reader just puts bytes in the slice and returns the number of bytes read.
> The writer then writes only the bytes that have been read.

Yeah, that's right.

> Also of note that we started with the intention of making it more efficient than allocating a new buffer. If we add zeroing out we pretty much lost all efficiency gains.

We *are* more efficient this way -- wrt. to memory churn, which is what we're trying to optimise here. Zeroing in a 4096-byte slice on my machine takes 70ns/op, it's negligible, so I think it doesn't hurt us to reset an array to avoid leaking data on a pool. It's simply more correct, and removes a potential source of bugs.

Reply:
> It's simply more correct

it's not, the buffer contents are irrelevant in its pattern of use.

> removes a potential source of bugs.

I don't think we remove any source of bugs here, we are just adding redundant (ugly) code!

Reply:
Still, it's proper housekeeping. But let's see what @Stebalien thinks!

Reply:
@Stebalien thinks it's unnecessary, so I yield, @vyzo! Good discussion.

Reply:
Note: If we _want_ to do this, we could try adding this feature to the buffer pool library directly (given some good real-world benchmarking). However, this is probably one of the less important places to clear the buffer given that we're handling encrypted data for a party that doesn't really trust us anyways.